File: cpp/array/WordSearch.cpp

class Solution {
public:
    bool result;
    int rows;
    int cols;
    std::vector<std::vector<char> > board;
    
public:
    Solution () {
        this->result = false;
    }
    
    bool exist(std::vector<std::vector<char> > &board, std::string word) {
        this->rows = int(board.size());
        this->cols = int(board[0].size());
        this->board = board;
        for (int i = 0; i < this->rows; i ++) {
            for (int j = 0; j < this->cols; j ++) {
                if (board[i][j] == word.at(0)) {
                    this->dfs(i, j, word.substr(1, word.length()-1));
                }
            }
        }
        return this->result;
    }
    
    void dfs(int i, int j, std::string word) {
        if (this->result) {
            return;
        }
        if (word.length() == 0) {
            this->result = true;
            return;
        }
        if (!this->result) {
            int up = i - 1;
            int down = i + 1;
            int left = j - 1;
            int right = j + 1;
            char temp = this->board[i][j];
            this->board[i][j] = '.';
            if (up >= 0 && this->board[up][j] == word.at(0)) {
                this->dfs(up, j, word.substr(1, word.length()-1));
            }
            if (down <= this->rows - 1 && this->board[down][j] == word.at(0)) {
                this->dfs(down, j, word.substr(1, word.length()-1));
            }
            if (left >= 0 && this->board[i][left] == word.at(0)) {
                this->dfs(i, left, word.substr(1, word.length()-1));
            }
            if (right <= this->cols - 1 && this->board[i][right] == word.at(0)) {
                this->dfs(i, right, word.substr(1, word.length()-1));
            }
            this->board[i][j] = temp;
        }
    }
};

File: archive/python/array/WordSearch.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def __init__(self):
        self.word = ""
        self.board = []
        self.rows = 0
        self.cols = 0
        self.result = False
        
    def exist(self, board, word):
        self.board = board
        self.rows = len(board)
        self.cols = len(board[0])
        self.word = word
        for i in xrange(self.rows):
            for j in xrange(self.cols):
                if board[i][j] == word[0]:
                    self.dfs(i, j, word[1:])
        return self.result
        
    def dfs(self, i, j, word):
        if self.result:
            return
        if len(word) == 0:
            self.result = True
            return True
        if not self.result:
            temp = self.board[i][j]
            self.board[i][j] = '.'
            up = i - 1
            down = i + 1
            left = j - 1
            right = j + 1
            if up >= 0 and self.board[up][j] == word[0]:
                self.dfs(up, j, word[1:])
            if down <= self.rows - 1 and self.board[down][j] == word[0]:
                self.dfs(down, j, word[1:])
            if left >= 0 and self.board[i][left] == word[0]:
                self.dfs(i, left, word[1:])
            if right <= self.cols - 1 and self.board[i][right] == word[0]:
                self.dfs(i, right, word[1:])
            self.board[i][j] = temp

if __name__ == "__main__":
    solution = Solution()

    print solution.exist(['a', 'a'], "aaa")
